add notes and questions for n28107 (#721)
bitcoin-core-review-club/website: _posts/2023-10-04-#28107.md

components: ["utils/log/libs"]
host: stickies-v
status: upcoming
commit: e74152d5b48bdca8dfc2217f35842dbb24b2746c
---

## Notes

- Since the activation of SegWit, transactions are identifiable by two types of identifiers: `txid` and `wtxid`. Both identifiers are a `uint256` resulting from a double `sha256` hash of serialized transaction data. The core [difference](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/primitives/transaction.h#L276-L280) is that `wtxid` includes the witness data, while `txid` does not.
 
- Type-safety is a concept in programming to help avoid type errors, enforcing strict correspondence between a variable and its type. It ensures that a program performs only legal operations on data, with validation occurring during compile-time. This proactive error detection and prevention lead to more reliable and maintainable software, minimizing unintended behavior and vulnerabilities. An example of an earlier type-safety improvement in Bitcoin Core is the incorporation of `std::chrono`, introduced in [#16908](https://github.com/bitcoin/bitcoin/pull/16908).

> - In the context, of transaction identifiers, type-safety is important because it is impossible to deduce the type difference from just the `uint256` value. With `GenTxid` type checking can be done, but it is more verbose and done at run-time instead of at compile-time. Additionally, transaction identifier types improve readability of code by being explicit about expected types instead of solely relying on clear variable naming.

Reviewer: ```suggestion
- In the context, of transaction identifiers, type-safety is important because it is impossible to deduce the type difference from just the `uint256` value. With `GenTxid` type checking can be done, but it is more verbose and done at run-time instead of at compile-time. Additionally, transaction identifier types improve the readability of code by being explicit about expected types instead of solely relying on clear variable naming.
```

- This PR aims to improve type safety by introducing the new [`transaction_identifier`](https://github.com/bitcoin-core-review-club/bitcoin/commit/28f9ff3d5416407d3c6d5374f6fc3c2767f4a7f3#diff-f3d91988bf7d98529d638fa829dc3fab695be002305a7614eaaeb283d4fb8101R18) type, with `Txid` and `Wtxid` typedefs for brevity.

> - There are instances where a variable could be either a `wtxid` or a `txid`. To navigate this, in some places like [`most_recent_block_txs`](https://github.com/bitcoin/bitcoin/blob/6619d6a8dca5a4d8e664a76526ac6bef3eb17831/src/net_processing.cpp#L1916-L1918), the `uint256` type is used. In other contexts such as [`CTxMemPool::info`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/txmempool.cpp#L862), the [`GenTxid`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/primitives/transaction.h#L425) type is adopted.

Reviewer 2: ```suggestion
- There are instances where `wtxid` and `txid` would be treated identically. To navigate this, in some places like [`most_recent_block_txs`](https://github.com/bitcoin/bitcoin/blob/6619d6a8dca5a4d8e664a76526ac6bef3eb17831/src/net_processing.cpp#L1916-L1918), the `uint256` type is used. In other contexts where `wtxid` and `txid` are treated differently but either would be okay, such as [`CTxMemPool::info`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/txmempool.cpp#L862), the [`GenTxid`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/primitives/transaction.h#L425) type is adopted.
```

## Questions

### Concept

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?

1. What does it mean for a transaction identifier to be type-safe? Why is that important or helpful? Are there any downsides?

> 1. Conceptually, when writing new code that requires referring to transactions, when should you use `txid` and when should you use `wtxid`?

Reviewer 2: Question idea: Can you point to any examples in the code where using one instead of the other could be very bad?

Reviewer 2: 1 point per example. Good students will come with lots of examples

Author: Yes, added!

> 1. In which concrete way(s) could using `transaction_identifier` instead of `uint256` help find existing bugs? On the other hand, could this change introduce new bugs?

Reviewer 2: ```suggestion
1. In which concrete way(s) could using `transaction_identifier` instead of `uint256` help find existing bugs or prevent the introduction of new ones? On the other hand, could this change introduce new bugs?
```

1. SegWit activated at block height 481,824. Do transactions in blocks < 481,824 also have a `wtxid`? For example, what about transaction with `txid` `dc01120a639283c11eeada6ce540178adfbbca7fe0d2f41076367cbc9e901822`?

2. Are there any places where a (w)txid is returned as a `uint256` where it would not make sense to return a `transaction_identifier` instead?

3. The [`GenTxid`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/primitives/transaction.h#L425) class already exists. How does it already enforce type correctness, and how is it different from the approach in this PR?

> 

Author: Added both, thank you!

4. Are there any places where `GenTxid` is used but using `transaction_identifier` would be better?

### Code Review

> 

Reviewer 2: Question idea: how is the type safety enforced? (Hint: what does the compiler say if you try to use a `Txid` in place of a `Wtxid` or vice versa?)

Author: I feel like this is already covered in

> Consider [`return m_wtxid_to_orphan_it.count(Wtxid::FromUint256(gtxid.GetHash()));`](https://github.com/bitcoin-core-review-club/bitcoin/commit/e74152d5b48bdca8dfc2217f35842dbb24b2746c#diff-7cfa3d46c37a17d80c78a37d1239b572a03095d50879ba610d87b702cd844f9dR177). Why is it not type-safe to allow `Wtxid` to be constructed by just passing a `uint256` to its constructor? Specifically, why doesn't `Wtxid{gtxid.GetHash()}` compile?

{:start="9"}
1. How is `transaction_identifier` able to subclass `uint256`, given that, in C++, integers are types and not classes?

1. Why does `transaction_identifier` subclass `uint256` instead of being a completely new type?

> 1. We define `Txid` and `Wtxid` as typedefs for `transaction_identifier<has_witness>`. Do you see any reason for code to directly use the `transaction_identifier` type? Should we forbid that, and if so, how?

Reviewer 2: I slightly prefer this kind of voice for talking about the PR code, but just a nit
```suggestion
1. `Txid` and `Wtxid` are defined as typedefs for `transaction_identifier<has_witness>`. Do you see any reason for code to directly use the `transaction_identifier` type? Should we forbid that, and if so, how?
```

Author: Agreed, done, for all other instances too.

## Questions

### Concept

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?

1. What does it mean for a transaction identifier to be type-safe? Why is that important or helpful? Are there any downsides?

1. Why is it better to enforce types at compile-time instead of at run-time?

1. Conceptually, when writing new code that requires referring to transactions, when should you use `txid` and when should you use `wtxid`? Can you point to any examples in the code where using one instead of the other could be very bad? (1 point per example)

1. In which concrete way(s) could using `transaction_identifier` instead of `uint256` help find existing bugs or prevent the introduction of new ones? On the other hand, could this change introduce new bugs?

> 1. SegWit activated at block height 481,824. Do transactions in blocks < 481,824 also have a `wtxid`? For example, what about transaction with `txid` `dc01120a639283c11eeada6ce540178adfbbca7fe0d2f41076367cbc9e901822`?

Reviewer: ```suggestion
1. SegWit activated at block height 481,824. Do transactions in blocks < 481,824 also have a `wtxid`? For example, what about the transaction with `txid` `dc01120a639283c11eeada6ce540178adfbbca7fe0d2f41076367cbc9e901822`?
```

1. Are there any places where a (w)txid is returned as a `uint256` where it would not make sense to return a `transaction_identifier` instead?

1. The [`GenTxid`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/primitives/transaction.h#L425) class already exists. How does it already enforce type correctness, and how is it different from the approach in this PR?

1. Are there any places where `GenTxid` is used but using `transaction_identifier` would be better?

### Code Review

{:start="10"}
1. How is `transaction_identifier` able to subclass `uint256`, given that, in C++, integers are types and not classes?

1. Why does `transaction_identifier` subclass `uint256` instead of being a completely new type?

1. `Txid` and `Wtxid` are typedefs for `transaction_identifier<has_witness>`. Do you see any reason for code to directly use the `transaction_identifier` type? Should that be forbidden, and if so, how?

> 1. [`transaction_identifier::Compare`](https://github.com/bitcoin-core-review-club/bitcoin/commit/28f9ff3d5416407d3c6d5374f6fc3c2767f4a7f3#diff-f3d91988bf7d98529d638fa829dc3fab695be002305a7614eaaeb283d4fb8101R39) is defined to return a `int`, yet `reinterpret_cast<const uint256&>` is used in the return statement. Is this correct, and if so, why? Why `reinterpret_cast` instead of another cast, such as `static_cast`?

Reviewer: ```suggestion
1. [`transaction_identifier::Compare`](https://github.com/bitcoin-core-review-club/bitcoin/commit/28f9ff3d5416407d3c6d5374f6fc3c2767f4a7f3#diff-f3d91988bf7d98529d638fa829dc3fab695be002305a7614eaaeb283d4fb8101R39) is defined to return an `int`, yet `reinterpret_cast<const uint256&>` is used in the return statement. Is this correct, and if so, why? Why `reinterpret_cast` instead of another cast, such as `static_cast`?
```

1. Is the [`has_witness`](https://github.com/bitcoin-core-review-club/bitcoin/commit/28f9ff3d5416407d3c6d5374f6fc3c2767f4a7f3#diff-f3d91988bf7d98529d638fa829dc3fab695be002305a7614eaaeb283d4fb8101R17) template parameter used for purposes other than type differentiation in `transaction_identifier`? What are the limitations of this approach? Do you see alternatives?

> 1. Why is `TxOrphanage` a good place to start introducing type-safe transaction ids?

Reviewer: ```suggestion
1. Why is `TxOrphanage` a good place to start introducing type-safe transaction IDs?
```
nit, not sure if we have a convention here
